stl/inc/barrier

```diff
@@ -0,0 +1,192 @@
+        _STL_VERIFY(_Update > 0, "Precondition: update > 0 (N4861 [thread.barrier.class]/12)");
+        // TRANSITION, GH-1133: should be memory_order_release
+        ptrdiff_t _Current = _Val._Myval2._Current.fetch_sub(_Update, memory_order_seq_cst) - _Update;
+        _STL_VERIFY(_Current >= 0, "Precondition: update is less than or equal to the expected count "
+                                   "for the current barrier phase (N4861 [thread.barrier.class]/12)");
+        if ((_Current & _Barrier_value_mask) == 0) {
+            // TRANSITION, GH-1133: should have this fence:
+            // atomic_thread_fence(memory_order_acquire);
+            _Completion(_Current);
+        }
+        return arrival_token{(_Current & _Barrier_arrival_token_mask) | reinterpret_cast<intptr_t>(this)};
+    }
+
+    void wait(arrival_token&& _Arrival) const noexcept /* strengthened */ {
+        _STL_VERIFY((_Arrival._Value & _Barrier_value_mask) == reinterpret_cast<intptr_t>(this),
+            "Preconditions: arrival is associated with the phase synchronization point for the current phase "
+            "or the immediately preceding phase of the same barrier object (N4861 [thread.barrier.class]/19)");
+        const ptrdiff_t _Arrival_value = _Arrival._Value & _Barrier_arrival_token_mask;
+        _Arrival._Value                = _Barrier_invalid_token;
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            const ptrdiff_t _Current = _Val._Myval2._Current.load(memory_order_seq_cst);
+            _STL_VERIFY(_Current >= 0, "Invariant counter >= 0, possibly caused by preconditions violation "
+                                       "(N4861 [thread.barrier.class]/12)");
+            if ((_Current & _Barrier_arrival_token_mask) != _Arrival_value) {
+                break;
+            }
+            _Val._Myval2._Current.wait(_Current, memory_order_relaxed);
+        }
+    }
+
+    void arrive_and_wait() noexcept /* strengthened */ {
+        // TRANSITION, GH-1133: should be memory_order_acq_rel
+        ptrdiff_t _Current =
+            _Val._Myval2._Current.fetch_sub(_Barrier_value_step, memory_order_seq_cst) - _Barrier_value_step;
+        const ptrdiff_t _Arrival = _Current & _Barrier_arrival_token_mask;
+        _STL_VERIFY(_Current >= 0, "Precondition: update is less than or equal to the expected count "
+                                   "for the current barrier phase (N4861 [thread.barrier.class]/12)");
+        if ((_Current & _Barrier_value_mask) == 0) {
+            _Completion(_Current);
+            return;
+        }
+
+        for (;;) {
+            _Val._Myval2._Current.wait(_Current, memory_order_relaxed);
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            _Current = _Val._Myval2._Current.load(memory_order_seq_cst);
+            _STL_VERIFY(_Current >= 0, "Invariant counter >= 0, possibly caused by preconditions violation "

```

Comment:
This seems incorrect. `arrive_and_drop` says that the precondition is that the current phase starts greater than 0, no statement at all on subsequent phases. So this seems valid:

```
barrier b{1};
b.arrive_and_drop();
```

which will leave _Current being 0.

Reply:
I don't see where it is incorrect. Counter reaches zero when callback is called.

You example is broken as you cannot drop twice and still wait in other thread.

Reply:
I think I read this as >0 rather than >= 0 for some reason.